core/domain/user_id_migration.py

```diff
@@ -0,0 +1,359 @@
+            model_class, old_user_id, new_user_id):
+        """Create new model with same values but new id and user_id."""
+        old_models = model_class.query(
+            model_class.get_user_id_migration_field() == old_user_id).fetch()
+        for old_model in old_models:
+            model_values = old_model.to_dict()
+            new_id = '%s.%s' % (new_user_id, old_model.id.split('.')[1])
+            model_values['id'] = new_id
+            model_values['user_id'] = new_user_id
+            model_class(**model_values).put(update_last_updated_time=False)
+            old_model.delete()
+
+    @staticmethod
+    def _change_model_with_one_user_id_field(
+            model_class, old_user_id, new_user_id):
+        """Replace field in model with new user id."""
+        migration_field = model_class.get_user_id_migration_field()
+        found_models = model_class.query(
+            migration_field == old_user_id).fetch()
+        for model in found_models:
+            model_values = model.to_dict()
+            model_values[migration_field._name] = new_user_id  # pylint: disable=protected-access
+            model.populate(**model_values)
+            model.put(update_last_updated_time=False)
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        """Return a list of datastore class references to map over."""
+        return [user_models.UserSettingsModel]
+
+    @staticmethod
+    def map(user_model):
+        """Implements the map function for this job."""
+        if user_model.id != user_model.gae_id:
+            return
+        old_user_id = user_model.id
+        new_user_id = user_models.UserSettingsModel.get_new_id('user')
+        for model_class in models.Registry.get_all_storage_model_classes():
+            if (model_class.get_user_id_migration_policy() ==
+                    base_models.USER_ID_MIGRATION_POLICY.NOT_APPLICABLE):
+                continue
+            elif (model_class.get_user_id_migration_policy() ==
+                  base_models.USER_ID_MIGRATION_POLICY.COPY):
+                UserIdMigrationJob._copy_model_with_new_id(
+                    model_class, old_user_id, new_user_id)
+            elif (model_class.get_user_id_migration_policy() ==
+                  base_models.USER_ID_MIGRATION_POLICY.COPY_PART):
+                UserIdMigrationJob._copy_model_with_new_id_and_user_id(
+                    model_class, old_user_id, new_user_id)
+            elif (model_class.get_user_id_migration_policy() ==
+                  base_models.USER_ID_MIGRATION_POLICY.ONE_FIELD):
+                UserIdMigrationJob._change_model_with_one_user_id_field(
+                    model_class, old_user_id, new_user_id)
+            elif (model_class.get_user_id_migration_policy() ==
+                  base_models.USER_ID_MIGRATION_POLICY.CUSTOM):
+                model_class.migrate_model(old_user_id, new_user_id)
+        yield ('SUCCESS', (old_user_id, new_user_id))
```

Comment:
Do you have ideas about where to add them? Should I add try-catch statements to some places and if we cathcsome error then yield failure? Because in the current code I don't see any obvious places where the migration should fail (compared to the verification jobs where we have specific places where we know that the job failed and failure is yielded).

Reply:
One example might be in _copy_model_with_new_id, when the old model doesn't exist -- at least we should get stats on that process.

Also we can't assume that everything in the datastore is "sound". Let's say for example that a field is "required" but it's got nothing in it, and we only discover it now because it's the first time we've retrieved-and-put a model in a long time, so an exception gets raised. What happens in cases like that? I think try/catch might be fragile since it can shadow the error, but just wanted to make sure things like that leave a visible trace that we'll notice.

Reply:
Added more `yield`s to some places if an unexpected error happens we will have to turn to the logs anyway.